### PR TITLE
New Unit Test: icon_state_sanity

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

Adds a new unit test ensuring atoms have a valid icon_state

## Why It's Good For The Game

Prevents coders from accidentally adding in invisible atoms.

## Changelog

Not player facing

### CODE REVIEW

File: code/modules/unit_tests/icon_state_sanity.dm [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/datum/unit_test/icon_state_sanity/Run()
```
Comment (Mothblocks) [inline code/modules/unit_tests/icon_state_sanity.dm:1]:
```suggestion
/// Documentation on what this does goes here
/datum/unit_test/icon_state_sanity

/datum/unit_test/icon_state_sanity/Run()
```

File: code/modules/unit_tests/icon_state_sanity.dm [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+	var/list/all_objs = typesof(/obj)
+	var/list/all_turfs = typesof(/turf)
+	var/list/all_mobs = typesof(/mob)
+	var/list/atoms_to_check = all_mobs | all_objs | all_turfs
```
Comment (Mothblocks) [inline code/modules/unit_tests/icon_state_sanity.dm:5]: `typesof(/atom) - typesof(/area)`? We want to check `/atom/movable` too, we have a few of those.

File: code/modules/unit_tests/icon_state_sanity.dm [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	for(var/atom_path in atoms_to_check)
+		if(isnull(atom_path))
```
Comment (Mothblocks) [inline code/modules/unit_tests/icon_state_sanity.dm:7]: How would it be?

Comment (Mothblocks) [inline code/modules/unit_tests/icon_state_sanity.dm:7]: How would it be?

File: code/modules/unit_tests/icon_state_sanity.dm [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+			continue
+		var/atom/path_ref = atom_path
```
Comment (Mothblocks) [inline code/modules/unit_tests/icon_state_sanity.dm:9]: Remove the above check and just use `as anything`

**Review by Mothblocks — CHANGES_REQUESTED** [review]
You didn't include it in unit_tests.dm

File: code/modules/unit_tests/icon_state_sanity.dm (hunk 2 above)
Comment (ZephyrTFA) [reply code/modules/unit_tests/icon_state_sanity.dm:5]: yea, that does it better actually

File: code/modules/unit_tests/icon_state_sanity.dm (hunk 3 above)
Comment (ZephyrTFA) [reply code/modules/unit_tests/icon_state_sanity.dm:4]: This is code from earlier where this kept happening for no real reason, it shouldnt

**Comment by ZephyrTFA** [comment]
> You didn't include it in unit_tests.dm

what does this mean exactly? Unit tests are run automatically using a subtypes iteration

**Comment by Mothblocks** [comment]
This isn't included in any dme, you need to put it in `unit_tests.dm`

**Comment by ZephyrTFA** [comment]
> This isn't included in any dme, you need to put it in `unit_tests.dm`

```
	var/list/tests_to_run = subtypesof(/datum/unit_test)
	var/list/focused_tests = list()
	for (var/_test_to_run in tests_to_run)
		var/datum/unit_test/test_to_run = _test_to_run
		if (initial(test_to_run.focus))
			focused_tests += test_to_run
```

**Comment by ZephyrTFA** [comment]
Alright, this is good to go, probably. Main issue is that there are 299 atoms that don't have a valid icon_state and as such will merging this will break CI horribly.

**Comment by LemonInTheDark** [comment]
You're gonna need to fix it then

**Comment by ZephyrTFA** [comment]
> 

I might be one of the worst people you can ask to do sprites. Best I can do is change them to the "shit is broke" icon state

**Comment by ZephyrTFA** [comment]
@LemonInTheDark

**Comment by LemonInTheDark** [comment]
I am unwilling to merge something that takes a lot of work and pushes it onto the mythical "spriters" in order to get ci working again.

It's hard I know, and you're not the most suited to it, but if it's merged as is we're basically introducing a bug that other people will be forced to fix. This is actually something we've done in the past with default purple and black sprites, and it burned a good few people out. Want to avoid that.

I don't think what you're looking at here is 300 objects with no icon state, I'm willing to bet it's more like 30 separate problems that manifest in 300 different typepaths.

Looking through here, I think most of these are less "sprite is missing, noooooo" errors, and more people or code misspelling icon states.

**Comment by ZephyrTFA** [comment]
Not stale

**Comment by LemonInTheDark** [comment]
Copium

**Pull request closed without merging**